pydantic/types.py

```diff
@@ -870,4 +872,60 @@ def __repr__(self) -> str:
             return 'NaiveDatetime'
 
 
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Encoded TYPES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+class EncoderProtocol(Protocol):
+    @classmethod
+    def decode(cls, data: bytes) -> bytes:
+        ...
+
+    @classmethod
+    def encode(cls, value: bytes) -> bytes:
+        ...
+
+
+class Base64Encoder(EncoderProtocol):
+    @classmethod
+    def decode(cls, data: bytes) -> bytes:
+        return base64.decodebytes(data)
+
+    @classmethod
+    def encode(cls, value: bytes) -> bytes:
+        return base64.encodebytes(value)
+
+
+@_dataclasses.dataclass
+class Encoded:
+    encoder: type[EncoderProtocol]
+
+    def __get_pydantic_core_schema__(
```

Comment:
We also need a `__get_pydantic_json_schema__`, I think it should just be the type of whatever is received. If Base64Bytes goes bytes to bytes, and Base64Str goes str to str then that keeps the schema simple (it can just be `{'type': 'string'}` or `{'type': 'bytes'}` as appropriate, I think). If it changes the input type from str to bytes or vice versa, don't worry too much about it for now.

Reply:
However there is `src/serializers/type_serializers/bytes.rs` in `pydantic_core` 🤔

Reply:
I've made it look like `{"type": "string", "format": <self.encoder.get_json_format()>}`